File: src/typegraph/deno/sdk/src/runtimes/kv.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import { Materializer, Runtime } from "./mod.ts";
import * as t from "../types.ts";
import { runtimes } from "../wit.ts";
import { Effect, KvMaterializer } from "../gen/typegraph_core.d.ts";

import { fx } from "../index.ts";

class KvOperationMat implements Materializer {
  operation: KvMaterializer;
  _id: number;
  constructor(id: number, operation: KvMaterializer) {
    this._id = id;
    this.operation = operation;
  }
}

export class KvRuntime extends Runtime {
  url: string;
  constructor(url: string) {
    const id = runtimes.registerKvRuntime({ url });
    super(id);
    this.url = url;
  }

  #operation(operation: KvMaterializer, effect: Effect) {
    const mad_id = runtimes.kvOperation({
      runtime: this._id,
      effect: effect,
    }, operation);

    return new KvOperationMat(mad_id, operation);
  }

  set() {
    const mat = this.#operation("set", fx.update(false));
    return t.func(
      t.struct({ "key": t.string(), "value": t.string() }),
      t.string(),
      mat,
    );
  }

  get() {
    const mat = this.#operation("get", fx.read());
    return t.func(t.struct({ "key": t.string() }), t.string(), mat);
  }

  delete() {
    const mat = this.#operation("delete", fx.delete_());
    return t.func(t.struct({ "key": t.string() }), t.integer(), mat);
  }

  keys() {
    const mat = this.#operation("keys", fx.read());
    return t.func(
      t.struct({ "filter": t.string().optional() }),
      t.list(t.string()),
      mat,
    );
  }

  values() {
    const mat = this.#operation("values", fx.read());
    return t.func(
      t.struct({ "filter": t.string().optional() }),
      t.list(t.string()),
      mat,
    );
  }
}

File: src/typegraph/rust/src/runtimes/kv.rs
use crate::{
    t::{self, TypeBuilder},
    wasm::{
        self,
        core::{MaterializerId, TypeId},
        runtimes::{BaseMaterializer, Effect, KvMaterializer, KvRuntimeData},
    },
    Result,
};

#[derive(Debug)]
pub struct KvRuntime {
    id: TypeId,
}

impl KvRuntime {
    pub fn new(url: impl ToString) -> Result<Self> {
        let data = KvRuntimeData {
            url: url.to_string(),
        };

        let id = wasm::with_runtimes(|r, s| r.call_register_kv_runtime(s, &data))?;

        Ok(Self { id })
    }

    pub fn set(&self) -> Result<TypeId> {
        let inp = t::r#struct()
            .prop("key", t::string())?
            .prop("value", t::string())?;
        let out = t::string();
        let mat = self.operation(KvMaterializer::Set, Effect::Update(false))?;

        t::func(inp, out, mat)?.build()
    }

    pub fn get(&self) -> Result<TypeId> {
        let inp = t::r#struct().prop("key", t::string())?;
        let out = t::string();
        let mat = self.operation(KvMaterializer::Get, Effect::Read)?;

        t::func(inp, out, mat)?.build()
    }

    pub fn delete(&self) -> Result<TypeId> {
        let inp = t::r#struct().prop("key", t::string())?;
        let out = t::integer();
        let mat = self.operation(KvMaterializer::Delete, Effect::Delete(false))?;

        t::func(inp, out, mat)?.build()
    }

    pub fn keys(&self) -> Result<TypeId> {
        let inp = t::r#struct().prop("filter", t::string().optional()?)?;
        let out = t::list(t::string())?;
        let mat = self.operation(KvMaterializer::Keys, Effect::Read)?;

        t::func(inp, out, mat)?.build()
    }

    pub fn values(&self) -> Result<TypeId> {
        let inp = t::r#struct().prop("filter", t::string().optional()?)?;
        let out = t::list(t::string())?;
        let mat = self.operation(KvMaterializer::Values, Effect::Read)?;

        t::func(inp, out, mat)?.build()
    }

    fn operation(&self, kind: KvMaterializer, effect: Effect) -> Result<MaterializerId> {
        let base = BaseMaterializer {
            runtime: self.id,
            effect,
        };

        wasm::with_runtimes(|r, s| r.call_kv_operation(s, base, kind))
    }
}
